staging/src/k8s.io/kube-apiextensions-server/pkg/controller/status/naming_controller.go

```diff
@@ -0,0 +1,312 @@
+				continue
+			}
+			if err := checkName(shortName, "", allResources); err != nil {
+				errs = append(errs, err)
+			}
+
+		}
+		if err := utilerrors.NewAggregate(errs); err != nil {
+			condition.Status = apiextensions.ConditionTrue
+			condition.Reason = "ShortNames"
+			condition.Message = err.Error()
+		} else {
+			newNames.ShortNames = requestedNames.ShortNames
+		}
+	}
+
+	if err := checkName(requestedNames.Kind, acceptedNames.Kind, allKinds); err != nil {
+		condition.Status = apiextensions.ConditionTrue
+		condition.Reason = "Kind"
+		condition.Message = err.Error()
+	} else {
+		newNames.Kind = requestedNames.Kind
+	}
+	if err := checkName(requestedNames.ListKind, acceptedNames.ListKind, allKinds); err != nil {
+		condition.Status = apiextensions.ConditionTrue
+		condition.Reason = "ListKind"
+		condition.Message = err.Error()
+	} else {
+		newNames.ListKind = requestedNames.ListKind
+	}
+
+	// if we haven't changed the condition, then our names must be good.
+	if condition.Status == apiextensions.ConditionUnknown {
+		condition.Status = apiextensions.ConditionFalse
+		condition.Reason = "Passed"
+		condition.Message = "no conflicts found"
+	}
+
+	return newNames, condition
+}
+
+func checkName(requestedName, acceptedName string, usedNames sets.String) error {
+	if requestedName == acceptedName {
+		return nil
+	}
+	if !usedNames.Has(requestedName) {
+		return nil
+	}
+
+	return fmt.Errorf("%q is already in use", requestedName)
+}
+
+func (c *NamingConditionController) sync(key string) error {
+	inCustomResourceDefinition, err := c.customResourceDefinitionLister.Get(key)
+	if apierrors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	acceptedNames, namingCondition := c.calculateNames(inCustomResourceDefinition)
+	if reflect.DeepEqual(inCustomResourceDefinition.Status.AcceptedNames, acceptedNames) &&
+		apiextensions.IsCustomResourceDefinitionEquivalent(
+			&namingCondition,
+			apiextensions.GetCustomResourceDefinitionCondition(inCustomResourceDefinition, apiextensions.NameConflict)) {
+		return nil
+	}
+
+	customResourceDefinition := &apiextensions.CustomResourceDefinition{}
+	if err := apiextensions.DeepCopy_apiextensions_CustomResourceDefinition(inCustomResourceDefinition, customResourceDefinition, cloner); err != nil {
+		return err
+	}
+
+	customResourceDefinition.Status.AcceptedNames = acceptedNames
+	apiextensions.SetCustomResourceDefinitionCondition(customResourceDefinition, namingCondition)
+
+	// if we're going to update our status, we may be releasing a name.  When this happens, we need to rekick everything in our group
+	list, err := c.customResourceDefinitionLister.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+	for _, curr := range list {
+		if curr.Spec.Group == customResourceDefinition.Spec.Group {
+			c.queue.Add(curr.Name)
+		}
+	}
+
+	_, err = c.customResourceDefinitionClient.CustomResourceDefinitions().UpdateStatus(customResourceDefinition)
```

Comment:
Do we know that the customResourceDefinitionLister is up-to-date immediately? I.e. do we see a consistent group state for the next item in the queue?